### PR TITLE
Move to libzim7

### LINKED ISSUE #69: Move to libzim7
Libzim7 has just been published and provides a lot of improvements. But the API has slightly changed, so we need to adapt our code. Probably also a good opportunity to clean a few things in the CI and the doc.

### LINKED ISSUE #79: Remove ability to compress using lzma
See https://github.com/openzim/mwoffliner/issues/1642 and https://github.com/openzim/libzim/issues/714

### PR DESCRIPTION
**Pull request by kelvinhammond**
Fixes #69
Fixes #79 

This is a work in progress. The test script is what I'm using to test things as I develop, I'll add proper tests later. Once I figure out the threading issues I'll remove / cleanup the code.

Let me know if you want the API changed. I went with accessor methods for some of the C++ class methods because that is the current javascript style. This code could be dried out with all the try catch stuff but then it ends up looking a bit unwieldy so its not dry for readability and flexibility in this way.

### CODE REVIEW

**Comment by kelson42** [comment]
@kelvinhammond Thank you for your PR. Should I try to build it and see how we could use it with mwoffliner?

**Comment by kelvinhammond** [comment]
> @kelvinhammond Thank you for your PR. Should I try to build it and see how we could use it with mwoffliner?

@kelson42 Its not quite ready yet, the creator doesn't work because there are some issues with accessing the javascript objects via a thread. I'm working on figuring that out. Single threaded might work but its missing some functions such as add illustration etc.

**Comment by kelson42** [comment]
@kelvinhammond FYI We have also made a new maintenance release of libzim, the version 7.1.0 https://download.openzim.org/release/libzim/. Nothing really important for you, but please use it if possible. Here is the changelog https://github.com/openzim/libzim/blob/master/ChangeLog

**Comment by kelson42** [comment]
@kelvinhammond Any news on this PR? Is there anything which can be done to help you to make it ready for review?

**Comment by kelvinhammond** [comment]
@kelson42 Sorry about the delay, I've been busy with life. I'll get started on this gain this weekend and hopefully finish this up.

**Comment by kelvinhammond** [comment]
Rebase from master and force pushed with lease to this branch.

File: .editorconfig [hunk 1]
```diff
@@ -15,3 +15,6 @@ indent_style = tab
 [*.{c,h,cpp,cpp,hpp}]
 indent_size = 4
 
+[*.{js,ts}]
```
Comment (kelvinhammond) [inline .editorconfig:18]: I added indents for javascript

File: CPPLINT.cfg [hunk 2]
```diff
@@ -13,6 +13,9 @@ filter=-build/namespaces_literals
 filter=-build/include_subdir
 filter=-legal/copyright
 
+# Allow runtime references
+filter=-runtime/references
```
Comment (kelvinhammond) [inline CPPLINT.cfg:17]: I changed cpplint's config to allow runtime references. A reference is safer than a pointer in this instance.

File: package.json [hunk 3]
```diff
@@ -1,13 +1,16 @@
 {
   "name": "@openzim/libzim",
+  "main": "dist/index.js",
+  "types": "dist/index.d.js",
   "version": "2.4.4",
   "description": "Libzim bindings for NodeJS",
   "scripts": {
     "clean": "rm -rf dist build/native/build",
     "tsc": "tsc",
```
Comment (kelvinhammond) [inline package.json:9]: We could probably remove this, no longer required, just use `src/index.js` directly.

File: package.json [hunk 4]
```diff
@@ -36,31 +38,31 @@
...
-    "ts-jest": "^25.4.0",
-    "ts-node": "^8.10.2",
+    "ts-jest": "^27.0.7",
+    "ts-node": "^10.4.0",
     "tsconfig-paths": "^3.11.0"
   },
   "devDependencies": {
     "nyc": "^15.1.0",
-    "typescript": "^3.9.10"
+    "typescript": "^4.4.4"
```
Comment (kelvinhammond) [inline package.json:65]: updated typescript

`4` is considered unlucky in many places in China :)

File: tsconfig.json [hunk 5]
```diff
@@ -2,16 +2,16 @@
...
     "lib": [
-      "es2015"
+      "es6"
     ],
     "declaration": true,
     "noImplicitAny": true,
     "outDir": "dist",
     "removeComments": false,
-    "target": "es5",
+    "target": "es6",
```
Comment (kelvinhammond) [inline tsconfig.json:11]: Wanted / needed `es6` so I upgraded it.

**Comment by kelvinhammond** [comment]
@kelson42 Can someone review this? I've basically rewritten the library again so that it supports the new libzim7 architecture. I'll updated mwoffliner after this has been merged and released.

**Comment by kelvinhammond** [comment]
Not sure how to fix the CI builds. Some of the builds use node 12 which is end of life this month. The ones that use node 14x+ pass. Can someone update those?

Codefactor is currently outdated and does not currently align with the CPPLint.cfg in the project for some reason.

**Comment by kelson42** [comment]
@kelvinhammond If Nodejs12 makes difficulties we can just remove it, see `.github/workflows/ci.yml`

**Comment by kelson42** [comment]
@ke

> @kelvinhammond If Nodejs12 makes difficulties we can just remove it, see `.github/workflows/ci.yml` I will, you will just need to rebase.

Done, and all the CI pass, so if something does not anymore with this PR. This is probably a regression.

BTW, we have meanwhile a libzim7.2.0 released at https://download.openzim.org/release/libzim/ and you can also anytime test with the cutting edge nightly at https://download.openzim.org/nightly/ (a few improvements have been done after your reported a problem and might be useful for you). If necessary, we could consider a patch release "just for you" getting the latest codebase for node-libzim 2.5.0

**Comment by kelvinhammond** [comment]
I'll make that PR soon then

**Comment by kelvinhammond** [comment]
Its ready

File: package.json (hunk 3 above)
Comment (kelson42) [reply package.json:9]: No opinion on this...

**Review by kelson42 — CHANGES_REQUESTED** [review]
@kelvinhammond So great to read we near us to the end of this PR! But before diving in details, my first point is: What about the macos CI, why does it fail?

**Comment by kelvinhammond** [comment]
@kelson42 It says node 14x but then I see this `gyp info using node@10.24.1` so I'm not sure what's wrong with the build

**Comment by kelson42** [comment]
First I have rebased on `master` (necessary anyway... and `master` passes)

**Comment by kelson42** [comment]
@kelvinhammond I believe I have achieved to fix the CI :)

**Comment by kelvinhammond** [comment]
Thank you. I tried it on a Mac and it worked, tests passed.

**Comment by kelvinhammond** [comment]
@kelson42 Codefactor is out of date and has not updated for some reason. The includes it complains about are all there.

I also updated the `CPPLINT.cfg` to allow for `#pragma once` which should be acceptable as a header guard.

Running `cpplint src/*.{h,cc}` gives
```
Done processing src/archive.h
Done processing src/blob.h
Done processing src/common.h
Done processing src/contentProvider.h
Done processing src/creator.h
Done processing src/entry.h
Done processing src/entryrange.h
Done processing src/item.h
Done processing src/module.cc
Done processing src/search.h
Done processing src/suggestion.h
Done processing src/writerItem.h
```
with no errors.

**Comment by kelson42** [comment]
> @kelson42 Codefactor is out of date and has not updated for some reason.

How do you see that? What is the current version, and what is the latest version? It's annoying if CodeFactor delivers too many false positive errors.

**Comment by kelvinhammond** [comment]
@kelson42 If you click on details next to the CodeFactor it shows this `https://www.codefactor.io/repository/github/openzim/node-libzim/pull/72` 

If you look at [blob.h](https://github.com/openzim/node-libzim/pull/72/files#diff-f28f98b9b15ae6a552a5b00f2b80c65a701a42b1e75f81aff149ac16e480b238) it includes `<string>` but CodeFactor complains about `Add #include <string> for string lines of code = 1`

It also is supposed to pull from the CPPLint.cfg which I've updated.

**Comment by kelson42** [comment]
@kelvinhammond I believe your latest commit is not proper…

**Comment by kelvinhammond** [comment]
@kelson42 I think I've fixed the previous commit

**Comment by kelson42** [comment]
@kelvinhammond Hmm, we still have a few things to fix looks like:
* jtest keeps reporting. Warnings
* Codecov seems to still ignore your configuration file
* Test seems to randomly segfault at setindex test (there is quite sure a bug here, probably in the libzim, we need a stablem reproduction case, maybe by running the test 100 times in a row?)

**Comment by kelvinhammond** [comment]
Working on figuring out a threading issue where libzim addItem adds an item to the queue and blocks until it can but blocks the main thread as a result

**Comment by kelson42** [comment]
@kelvinhammond Hope you work with the nightly of libzim. A new libzim release should be done today or tomorrow.

**Comment by kelson42** [comment]
@kelvinhammond libzim 7.2.1 is now available at http://download.openzim.org/release/libzim/

**Comment by kelvinhammond** [comment]
@kelson42 version updated.

**Comment by kelson42** [comment]
@kelvinhammond Thx, problems with macOS in the CI are normal, we are handling them.

**Comment by kelvinhammond** [comment]
@kelson42 I updated the test-mem-leak script but now I'm seeing a new issue with creating a large zim, I'll get back to you once I figure it out and push a fix.

**Review by kelson42 — APPROVED** [review]
From a user perspective it looks good. I don't see DEPRECATED warnings during the compilation too. @mgautierfr Please have a look to the code.

**Comment by kelvinhammond** [comment]
Issue I'm working on here: https://github.com/nodejs/node-addon-api/issues/1174

**Comment by kelson42** [comment]
CI Is back green for macOS

**Comment by kelvinhammond** [comment]
@kelson42 I'll debug this a bit more tomorrow but currently we're waiting on the nodejs node-addon-api team to see if they can fix the Segmentation fault issue(s).

Otherwise this is ready to go, I'll probably start on the mwoffliner changes while waiting on a fix for that, the api for this repo _should_ _not_ change that much.

**Comment by kelson42** [comment]
@kelvinhammond Sounds good! We have released a new libzim 7.2.2, you can just use it,  ut should not change anything for you.

**Comment by kelson42** [comment]
> I also updated the `CPPLINT.cfg` to allow for `#pragma once` which should be acceptable as a header guard.

It seems to be a weakness in cpplint! https://github.com/cpplint/cpplint/issues/65

**Comment by kelvinhammond** [comment]
Still waiting on the node-addon-api team for a fix, not sure when they'll have something. Any ideas on what to do next?

**Comment by kelson42** [comment]
@kelvinhammond What is exactly the impact of the nodejs "bug" in mwoffliner? Is that aporadic, systematic?

**Comment by kelvinhammond** [comment]
@kelson42 The issue happens on pretty much every run, it is reproducible.

**Comment by kelson42** [comment]
@kelvinhammond Would that be wise/workable to start a PR on mwoffliner to adapt the code to new libzim7?

**Comment by kelvinhammond** [comment]
I've already started looking at mwoffliner

**Comment by kelson42** [comment]
@kelvinhammond What about merging this PR and opening a dedicated ticket for the crash scenario? This would make it easier to move on I believe.

**Comment by kelvinhammond** [comment]
The test-mem-leak test does not pass because of this and also will not work with larger zim files because of this bug.

**Comment by kelvinhammond** [comment]
@kelson42 @mgautierfr Does this seem normal for speed on an ssd laptop?
> |----------| 6438/1000000 0% [elapsed: 02:03, left: 34327777:26:04, 52.10 iters/s]
Default settings, clustersize 2048, indexing on with "en";

**Comment by kelson42** [comment]
@kelvinhammond I'm not sure to understand the question. Where this log comes from. What has been done to get it?

**Comment by kelvinhammond** [comment]
> @kelvinhammond I'm not sure to understand the question. Where this log comes from. What has been done to get it?

I was testing libzim calls and it was doing about 52 iterations per second which seemed kinda slow. I wanted to know if this was normal or if there was a bottleneck somewhere.

**Comment by kelson42** [comment]
@kelvinhammond Unfortunately I have no opinion on this

**Comment by mgautierfr** [comment]
Iterations on what ?
Reading entries ? You read only 52 entries per second ?
Or writing ?

In any case, it seems a bit slow yes.

**Comment by kelvinhammond** [comment]
I seem to be stuck with the nodejs bug at the moment, not sure what I should do from here.

https://github.com/nodejs/node-addon-api/issues/1174

**Comment by kelson42** [comment]
@kelvinhammond Yes, not easy. I would recommend to push on the mwoffliner front in the meantime, so we will be ready there too when this bug will be fixed.

**Comment by kelson42** [comment]
@kelvinhammond libzim8.0.0 has been published.

**Comment by kelvinhammond** [comment]
@kelson42
- Rebased with origin/master
- Updated dependencies
- Updated to libzim 8.0.0
- Removed lzma compression.
- I've updated the code so that file and string items work, the custom item however does not work with the makeLargeZim.js . We can fix that later (or nodejs can fix it). The custom item does work but it crashes after about 10k or so with lots of threads.

I think this PR is ready to go, what do we need to do next?

**Comment by kelson42** [comment]
@kelvinhammond Agree with you, but before merging, can you please open a ticket with a clear description of the problem/regression please?

**Comment by kelvinhammond** [comment]
@kelson42 issue here https://github.com/openzim/node-libzim/issues/80

**Review by kelson42 — APPROVED** [review]
@kelvinhammond Thank you so much for the work on this big and crucial PR
